README.md
# Meteor FreeBSD

This is the latest 1.10.2 version of Meteor, always updated to use the latest NodeJS available in FreeBSD ports. I'm using this to develop Meteor apps on my FreeBSD laptop and I do my best to keep it up to date.


History.md

## v1.10.2, 2020-04-21

### Breaking changes

* The `babel-compiler` package, used by both `ecmascript` and
  `typescript`, no longer supports stripping [Flow](https://flow.org/)
  type annotations by default, which may be a breaking change if your
  application (or Meteor package) relied on Flow syntax.

### Migration steps

* If you still need Babel's Flow plugins, you can install them with npm
  and then enable them with a custom `.babelrc` file in your application's
  (or package's) root directory:
  ```json
  {
    "plugins": [
      "@babel/plugin-syntax-flow",
      "@babel/plugin-transform-flow-strip-types"
    ]
  }
  ```

### Changes

* Node.js has been updated to version
[13.3.0](https://nodejs.org/en/blog/release/v13.3.0/).
* Adds support to override MongoDB options via Meteor settings. Code PR 
[#10976](https://github.com/meteor/meteor/pull/10976), Docs PR 
[#662](https://github.com/meteor/docs/pull/662)

* The `meteor-babel` npm package has been updated to version 7.9.0.

* The `typescript` npm package has been updated to version 3.8.3.
  
* To pass Node command line flags to the server node instance,
  now it is recommended to use `SERVER_NODE_OPTIONS` instead of `NODE_OPTIONS`.
  Since Meteor 0.5.3, Meteor allowed to pass node command line flags via the  `NODE_OPTIONS`
  environment variable.
  However, since Node version 8 / Meteor 1.6 this has become a default node
  envar with the same behavior. The side effect is that this now also affects
  Meteor tool. The command line parameters could already be set separately
  via the `TOOL_NODE_FLAGS` envar. This is now also possible (again) for the server.

* The version of MongoDB used by Meteor in development has been updated from
  4.2.1 to 4.2.5.
  [PR #11020](https://github.com/meteor/meteor/pull/11020)

## v1.10.1, 2020-03-12

### Breaking changes

* Cordova has been updated from version 7 to 9. We recommend that you test
  your features that are taking advantage of Cordova plugins to be sure
  they are still working as expected.
  
  * WKWebViewOnly is set by default now as true so if you are relying on 
  UIWebView or plugins that are using UIWebView APIs you probably want to
  set it as false, you can do this by calling 
  `App.setPreference('WKWebViewOnly', false);` in your mobile-config.js. But we 
  don't recommend turning this into false because 
  [Apple have said](https://developer.apple.com/news/?id=12232019b) they are 
  going to reject apps using UIWebView.
  
* Because MongoDB since 3.4 no longer supports 32-bit Windows, Meteor 1.10 has
  also dropped support for 32-bit Windows. In other words, Meteor 1.10 supports
  64-bit Mac, Windows 64-bit, and Linux 64-bit.

### Migration Steps
* If you get `Unexpected mongo exit code 62. Restarting.` when starting your local
  MongoDB, you can either reset your project (`meteor reset`)
  (if you don't care about your local data)
  or you will need to update the feature compatibility version of your local MongoDB:
  
    1. Downgrade your app to earlier version of Meteor `meteor update --release 1.9.2`
    2. Start your application
    3. While your application is running open a new terminal window, navigate to the
       app directory and open `mongo` shell: `meteor mongo`
    4. Use: `db.adminCommand({ getParameter: 1, featureCompatibilityVersion: 1 })` to
       check the current feature compatibility.
    5. If the returned version is less than 4.0 update like this:
       `db.adminCommand({ setFeatureCompatibilityVersion: "4.2" })`
    6. You can now stop your app and update to Meteor 1.10.
    
    For more information about this, check out [MongoDB documentation](https://docs.mongodb.com/manual/release-notes/4.2-upgrade-standalone/).

### Changes

* The version of MongoDB used by Meteor in development has been updated
  from 4.0.6 to 4.2.1, and the `mongodb` driver package has been updated
  from 3.2.7 to 3.5.4, thanks to [@klaussner](https://github.com/klaussner).
  [Feature #361](https://github.com/meteor/meteor-feature-requests/issues/361)
  [PR #10723](https://github.com/meteor/meteor/pull/10723)

* The `npm` command-line tool used by the `meteor npm` command (and by
  Meteor internally) has been updated to version 6.14.0, and our
  [fork](https://github.com/meteor/pacote/tree/v9.5.12-meteor) of its
  `pacote` dependency has been updated to version 9.5.12.

* Cordova was updated from version 7 to 9
  * cordova-lib from 7.1.0 to 9.0.1 [release notes](https://github.com/apache/cordova-lib/blob/master/RELEASENOTES.md)
  * cordova-common from 2.1.1 to 3.2.1 [release notes](https://github.com/apache/cordova-common/blob/master/RELEASENOTES.md)
  * cordova-android from 7.1.4 to 8.1.0 [release notes](https://github.com/apache/cordova-android/blob/master/RELEASENOTES.md)
  * cordova-ios from 4.5.5 to 5.1.1 [release notes](https://github.com/apache/cordova-ios/blob/master/RELEASENOTES.md)
  * cordova-plugin-wkwebview-engine from 1.1.4 to 1.2.1 [release notes](https://github.com/apache/cordova-plugin-wkwebview-engine/blob/master/RELEASENOTES.md#121-jul-20-2019)
  * cordova-plugin-whitelist from 1.3.3 to 1.3.4 [release notes](https://github.com/apache/cordova-plugin-whitelist/blob/master/RELEASENOTES.md#134-jun-19-2019)
  * cordova-plugin-splashscreen (included by mobile-experience > launch-screen) 
  from 4.1.0 to 5.0.3 [release notes](https://github.com/apache/cordova-plugin-splashscreen/blob/master/RELEASENOTES.md#503-may-09-2019)
  * cordova-plugin-statusbar (included by mobile-experience > mobile-status-bar) 
  from 2.3.0 to 2.4.3 [release notes](https://github.com/apache/cordova-plugin-statusbar/blob/master/RELEASENOTES.md#243-jun-19-2019)
  * On iOS WKWebViewOnly is set by default now as true.
  * On iOS the Swift version is now set by default to `5` this change can make
  your app to produce some warnings if your plugins are using old Swift code. 
  You can override the Swift version using 
  `App.setPreference('SwiftVersion', 4.2);` but we don't recommend that.

scripts/freebsd/bootstrap.js
var main = require('../../tools/cli/main.js');
var _ = require('underscore');
var files = require('../../tools/fs/files');
var buildmessage = require('../../tools/utils/buildmessage.js');
var config = require('../../tools/meteor-services/config.js');
var utils = require('../../tools/utils/utils.js');
var catalog = require('../../tools/packaging/catalog/catalog.js');
var catalogRemote = require('../../tools/packaging/catalog/catalog-remote.js');
var isopack = require('../../tools/isobuild/isopack.js');
var Console = require('../../tools/console/console.js').Console;
var catalogUtils = require('../../tools/packaging/catalog/catalog-utils.js');

var packageMapModule = require('../../tools/packaging/package-map.js');
var packageClient = require('../../tools/packaging/package-client.js');
var tropohouse = require('../../tools/packaging/tropohouse.js');

const argv = require('minimist')(process.argv);

console.log(argv);

const makeBootstrapTarballsFreeBSD = options => {
  var releaseNameAndVersion = options.args[0];

  // We get this as an argument, so it is an OS path. Make it a standard path.
  var outputDirectory = files.convertToStandardPath(options.args[1]);

  var trackAndVersion = catalogUtils.splitReleaseName(releaseNameAndVersion);
  var releaseTrack = trackAndVersion[0];
  var releaseVersion = trackAndVersion[1];

  var releaseRecord = catalog.official.getReleaseVersion(
    releaseTrack, releaseVersion);
  if (!releaseRecord) {
    // XXX this could also mean package unknown.
    Console.error('Release unknown: ' + releaseNameAndVersion + '');
    return 1;
  }

  var toolPackageVersion = releaseRecord.tool &&
    utils.parsePackageAndVersion(releaseRecord.tool);
  if (!toolPackageVersion) {
    throw new Error("bad tool in release: " + releaseRecord.tool);
  }
  var toolPackage = toolPackageVersion.package;
  var toolVersion = toolPackageVersion.version;

  var toolPkgBuilds = catalog.official.getAllBuilds(
    toolPackage, toolVersion);
  if (!toolPkgBuilds) {
    // XXX this could also mean package unknown.
    Console.error('Tool version unknown: ' + releaseRecord.tool);
    return 1;
  }
  if (!toolPkgBuilds.length) {
    Console.error('Tool version has no builds: ' + releaseRecord.tool);
    return 1;
  }

  // XXX check to make sure this is the three arches that we want? it's easier
  // during 0.9.0 development to allow it to just decide "ok, i just want to
  // build the OSX tarball" though.
  var buildArches = _.pluck(toolPkgBuilds, 'buildArchitectures');
  var osArches = _.map(buildArches, function (buildArch) {
    var subArches = buildArch.split('+');
    var osArches = _.filter(subArches, function (subArch) {
      return subArch.substr(0, 3) === 'os.';
    });
    if (osArches.length !== 1) {
      throw Error("build architecture " + buildArch + "  lacks unique os.*");
    }
    return osArches[0];
  });

  if (options['target-arch']) {
    // check if the passed arch is in the list
    var arch = options['target-arch'];
    if (!_.contains(osArches, arch)) {
      throw new Error(
        arch + ": the arch is not available for the release. Available arches: "
        + osArches.join(', '));
    }

    // build only for the selected arch
    osArches = [arch];
  }

  Console.error(
    'Building bootstrap tarballs for architectures ' + osArches.join(', '));

  // Before downloading anything, check that the catalog contains everything we
  // need for the OSes that the tool is built for.
  main.captureAndExit("=> Errors finding builds:", function () {
    _.each(osArches, function (osArch) {
      _.each(releaseRecord.packages, function (pkgVersion, pkgName) {
        buildmessage.enterJob({
          title: "looking up " + pkgName + "@" + pkgVersion + " on " + osArch
        }, function () {
          if (!catalog.official.getBuildsForArches(pkgName, pkgVersion, [osArch])) {
            buildmessage.error("missing build of " + pkgName + "@" + pkgVersion +
              " for " + osArch);
          }
        });
      });
    });
  });

  files.mkdir_p(outputDirectory);

  // Get a copy of the data.json.
  var dataTmpdir = files.mkdtemp();
  var tmpDataFile = files.pathJoin(dataTmpdir, 'packages.data.db');

  var tmpCatalog = new catalogRemote.RemoteCatalog();
  tmpCatalog.initialize({
    packageStorage: tmpDataFile
  });
  try {
    packageClient.updateServerPackageData(tmpCatalog, null);
  } catch (err) {
    packageClient.handlePackageServerConnectionError(err);
    return 2;
  }

  // Since we're making bootstrap tarballs, we intend to recommend this release,
  // so we should ensure that once it is downloaded, it knows it is recommended
  // rather than having a little identity crisis and thinking that a past
  // release is the latest recommended until it manages to sync.
  tmpCatalog.forceRecommendRelease(releaseTrack, releaseVersion);
  tmpCatalog.closePermanently();
  if (files.exists(tmpDataFile + '-wal')) {
    throw Error("Write-ahead log still exists for " + tmpDataFile
      + " so the data file will be incomplete!");
  }

  var packageMap =
    packageMapModule.PackageMap.fromReleaseVersion(releaseRecord);

  _.each(osArches, function (osArch) {
    var tmpdir = files.mkdtemp();
    Console.info("Building tarball for " + osArch);

    // when building for Windows architectures, build Windows-specific
    // tropohouse and bootstrap tarball
    var targetPlatform;
    if (/win/i.test(osArch)) {
      targetPlatform = "win32";
    }

    // We're going to build and tar up a tropohouse in a temporary directory.
    var tmpTropo = new tropohouse.Tropohouse(
      files.pathJoin(tmpdir, '.meteor'),
      { platform: targetPlatform });

    main.captureAndExit(
      "=> Errors downloading packages for " + osArch + ":",
      function () {
        tmpTropo.downloadPackagesMissingFromMap(packageMap, {
          serverArchitectures: [osArch]
        });
      }
    );

    // Install the sqlite DB file we synced earlier. We have previously
    // confirmed that the "-wal" file (which could contain extra log entries
    // that haven't been flushed to the main file yet) doesn't exist, so we
    // don't have to copy it.
    files.copyFile(tmpDataFile, config.getPackageStorage({
      root: tmpTropo.root
    }));

    // Create the top-level 'meteor' symlink, which links to the latest tool's
    // meteor shell script.
    var toolIsopackPath =
      tmpTropo.packagePath(toolPackage, toolVersion);
    var toolIsopack = new isopack.Isopack;
    toolIsopack.initFromPath(toolPackage, toolIsopackPath);
    var toolRecord = _.findWhere(toolIsopack.toolsOnDisk, { arch: osArch });
    if (!toolRecord) {
      throw Error("missing tool for " + osArch);
    }

    tmpTropo.linkToLatestMeteor(files.pathJoin(
      tmpTropo.packagePath(toolPackage, toolVersion, true),
      toolRecord.path,
      'meteor'));

    if (options.unpacked) {
      files.cp_r(tmpTropo.root, outputDirectory);
    } else {
      files.createTarball(
        tmpTropo.root,
        files.pathJoin(outputDirectory,
          'meteor-bootstrap-' + osArch + '.tar.gz'));
    }
  });

  return 0;
}

scripts/build-dev-bundle-common.sh
#!/usr/bin/env bash

set -e
set -u

UNAME=$(uname)
ARCH=$(uname -m)
NODE_VERSION=13.3.0
MONGO_VERSION_64BIT=4.2.5
MONGO_VERSION_32BIT=3.2.22
NPM_VERSION=6.14.0

# If we built Node from source on Jenkins, this is the build number.
NODE_BUILD_NUMBER=

if [ "$UNAME" == "Linux" ] ; then
    if [ "$ARCH" != "i686" -a "$ARCH" != "x86_64" ] ; then
        echo "Unsupported architecture: $ARCH"
        echo "Meteor only supports i686 and x86_64 for now."
        exit 1
    fi

    OS="linux"

        stripBinary() {
        strip --remove-section=.comment --remove-section=.note $1
    }
elif [ "$UNAME" == "FreeBSD" ] ; then
    if [ "$ARCH" != "i386" -a "$ARCH" != "amd64" ] ; then
        echo "Unsupported architecture: $ARCH"
        echo "Meteor only supports i386 and amd64 for now."
        exit 1
    fi

    OS="freebsd"

    stripBinary() {
        strip --remove-section=.comment --remove-section=.note $1
    }
elif [ "$UNAME" == "Darwin" ] ; then
    SYSCTL_64BIT=$(sysctl -n hw.cpu64bit_capable 2>/dev/null || echo 0)
    if [ "$ARCH" == "i386" -a "1" != "$SYSCTL_64BIT" ] ; then
        # some older macos returns i386 but can run 64 bit binaries.
        # Probably should distribute binaries built on these machines,
        # but it should be OK for users to run.
        ARCH="x86_64"
    fi

    if [ "$ARCH" != "x86_64" ] ; then
        echo "Unsupported architecture: $ARCH"
        echo "Meteor only supports x86_64 for now."
        exit 1
    fi

    OS="macos"

    # We don't strip on Mac because we don't know a safe command. (Can't strip
    # too much because we do need node to be able to load objects like
    # fibers.node.)
    stripBinary() {
        true
    }
else
    echo "This OS not yet supported"
    exit 1
fi

PLATFORM="${UNAME}_${ARCH}"

if [ "$UNAME" == "Linux" ]
then
    if [ "$ARCH" == "i686" ]
    then
        NODE_TGZ="node-v${NODE_VERSION}-linux-x86.tar.gz"
    elif [ "$ARCH" == "x86_64" ]
    then
        NODE_TGZ="node-v${NODE_VERSION}-linux-x64.tar.gz"
    else
        echo "Unknown architecture: $UNAME $ARCH"
        exit 1
    fi
elif [ "$UNAME" == "FreeBSD" ]
then
    if [ "$ARCH" == "i386" ]
    then
        NODE_TGZ="node-v${NODE_VERSION}-freebsd-x86.tar.gz"
    elif [ "$ARCH" == "amd64" ]
    then
        NODE_TGZ="node-v${NODE_VERSION}-freebsd-x64.tar.gz"
    else
        echo "Unknown architecture: $UNAME $ARCH"
        exit 1
    fi
elif [ "$UNAME" == "Darwin" ]
then
    NODE_TGZ="node-v${NODE_VERSION}-darwin-x64.tar.gz"
else
    echo "Unknown architecture: $UNAME $ARCH"
    exit 1
fi

SCRIPTS_DIR=$(dirname $0)
cd "$SCRIPTS_DIR/.."
CHECKOUT_DIR=$(pwd)

DIR=$(mktemp -d -t generate-dev-bundle-XXXXXXXX)
trap 'rm -rf "$DIR" >/dev/null 2>&1' 0

cd "$DIR"
chmod 755 .
umask 022
mkdir build
cd build


scripts/dev-bundle-tool-package.js
// This file contains a package.json for the dependencies of the command-line
// tool.

// We put this in a JS file so that it can contain comments. It is processed
// into a package.json file by generate-dev-bundle.sh.

var packageJson = {
  name: "meteor-dev-bundle-tool",
  private: true,
  dependencies: {
    // Explicit dependency because we are replacing it with a bundled version
    // and we want to make sure there are no dependencies on a higher version
    npm: "6.13.0",
    pacote: "https://github.com/meteor/pacote/tarball/a1853763676463c458aa8ce7b91c20205c44ae6a",
    "node-gyp": "5.0.1",
    "node-pre-gyp": "0.13.0",
    typescript: "3.7.2",
    "meteor-babel": "7.7.0",
    // Keep the versions of these packages consistent with the versions
    // found in dev-bundle-server-package.js.
    "meteor-promise": "0.8.7",
    fibers: "4.0.2",
    reify: "0.20.12",
    // So that Babel can emit require("@babel/runtime/helpers/...") calls.
    "@babel/runtime": "7.7.0",
    // For backwards compatibility with isopackets that still depend on
    // babel-runtime rather than @babel/runtime.
    "babel-runtime": "7.0.0-beta.3",
    "@types/underscore": "1.9.2",
    underscore: "1.9.1",
    "source-map-support": "https://github.com/meteor/node-source-map-support/tarball/1912478769d76e5df4c365e147f25896aee6375e",
    "@types/semver": "5.4.0",
    semver: "5.4.1",
    request: "2.88.0",
    uuid: "3.3.2",
    "graceful-fs": "4.2.2",
    fstream: "https://github.com/meteor/fstream/tarball/cf4ea6c175355cec7bee38311e170d08c4078a5d",
    tar: "2.2.1",
    // Fork of kexec@3.0.0 with my Node.js 12 compatibility PR
    // https://github.com/jprichardson/node-kexec/pull/37 applied.
    kexec: "https://github.com/meteor/node-kexec/tarball/f29f54037c7db6ad29e1781463b182e5929215a0",
    "source-map": "0.5.7",
    chalk: "0.5.1",
    sqlite3: "4.1.0",
    "http-proxy": "1.16.2",
    "is-reachable": "3.1.0",
    "wordwrap": "0.0.2",
    "moment": "2.20.1",
    "rimraf": "2.6.2",
    "glob": "7.1.2",
    ignore: "3.3.7",
    // XXX: When we update this, see if it fixes this Github issue:
    // https://github.com/jgm/CommonMark/issues/276 . If it does, remove the
    // workaround from the tool.
    "commonmark": "0.15.0",
    escope: "3.6.0",
    split2: "2.2.0",
    multipipe: "2.0.1",
    pathwatcher: "8.1.0",
    // The @wry/context package version must be compatible with the
    // version constraint imposed by optimism/package.json.
    optimism: "0.11.3",
    "@wry/context": "0.4.4",
    'lru-cache': '4.1.3'
  }
};

if (process.platform === 'win32') {
  // Remove dependencies that do not work on Windows
  delete packageJson.dependencies.netroute;
  delete packageJson.dependencies.kexec;
}

process.stdout.write(JSON.stringify(packageJson, null, 2) + '\n');


scripts/dev-bundle-server-package.js
// This file contains a package.json for the dependencies of the *BUNDLED
// SERVER* (not the command-line tool).

// We put this in a JS file so that it can contain comments. It is processed
// into a package.json file by generate-dev-bundle.sh.

var packageJson = {
  name: "meteor-dev-bundle",
  private: true,
  dependencies: {
    // Keep the versions of these packages consistent with the versions
    // found in dev-bundle-tool-package.js.
    fibers: "4.0.3",
    "meteor-promise": "0.8.7",
    promise: "8.0.3",
    reify: "0.20.12",
    "@babel/parser": "7.9.4",
    "@types/underscore": "1.9.2",
    underscore: "1.9.1",
    "source-map-support": "https://github.com/meteor/node-source-map-support/tarball/1912478769d76e5df4c365e147f25896aee6375e",
    "@types/semver": "5.4.0",
    semver: "5.4.1"
  },
  // These are only used in dev mode (by shell.js) so end-users can avoid
  // needing to install them if they use `npm install --production`.
  devDependencies: {
    split2: "2.1.1",
    multipipe: "1.0.2",
    chalk: "0.5.1"
  }
};

process.stdout.write(JSON.stringify(packageJson, null, 2) + '\n');


tools/utils/archinfo.ts
import { max } from 'underscore';
import os from 'os';
const utils = require('./utils');

/* Meteor's current architecture scheme defines the following virtual
 * machine types, which are defined by specifying what is promised by
 * the host environment:
 *
 * browser.w3c
 *   A web browser compliant with modern standards. This is
 *   intentionally a broad definition. In the coming years, as web
 *   standards evolve, we will likely tighten it up.
 *
 * browser.ie[678]
 *   Old versions of Internet Explorer (not sure yet exactly which
 *   versions to distinguish -- maybe 6 and 8?)
 *
 * os.linux.x86_64
 *   Linux on Intel x86 architecture. x86_64 means a system that can
 *   run 64-bit images, furnished with 64-bit builds of shared
 *   libraries (there is no guarantee that 32-bit builds of shared
 *   libraries will be available). x86_32 means a system that can run
 *   32-bit images, furnished with 32-bit builds of shared libraries.
 *   Additionally, if a package contains shared libraries (for use by
 *   other packages), then if the package is built for x86_64, it
 *   should contain a 64-bit version of the library, and likewise for
 *   32-bit.
 *
 *   Operationally speaking, if you worked at it, under this
 *   definition it would be possible to build a Linux system that can
 *   run both x86_64 and x86_32 images (eg, by using a 64-bit kernel
 *   and making sure that both versions of all relevant libraries were
 *   installed). But we require such a host to decide whether it is
 *   x86_64 or x86_32, and stick with it. You can't load a combination
 *   of packages from each and expect them to work together, because
 *   if they contain shared libraries they all need to have the same
 *   architecture.
 *
 *   Basically the punchline is: if you installed the 32-bit version
 *   of Ubuntu, you've got a os.linux.x86_32 system and you will
 *   use exclusively os.linux.x86_32 packages, and likewise
 *   64-bit. They are two parallel universes and which one you're in
 *   is determined by which version of Red Hat or Ubuntu you
 *   installed.
 *
 * os.osx.x86_64
 *   OS X (technically speaking, Darwin) on Intel x86 architecture,
 *   with a kernel capable of loading 64-bit images, and 64-bit builds
 *   of shared libraries available.  If a os.osx.x86_64 package
 *   contains a shared library, it is only required to provide a
 *   64-bit version of the library (it is not required to provide a
 *   fat binary with both 32-bit and 64-bit builds).
 *
 *   Note that in modern Darwin, both the 32 and 64 bit versions of
 *   the kernel can load 64-bit images, and the Apple-supplied shared
 *   libraries are fat binaries that include both 32-bit and 64-bit
 *   builds in a single file. So it is technically fine (but
 *   discouraged) for a os.osx.x86_64 to include a 32-bit
 *   executable, if it only uses the system's shared libraries, but
 *   you'll run into problems if shared libraries from other packages
 *   are used.
 *
 *   There is no os.osx.x86_32. Our experience is that such
 *   hardware is virtually extinct. Meteor has never supported it and
 *   nobody has asked for it.
 *
 * os.windows.x86_64
 *   Once, on the far side of yesterday, there was not a 64-bit
 *   build of Meteor for Windows, due to the belief that Node didn't
 *   take (enough?) advantage of a 64-bit platform.  As time has passed,
 *   and as V8 engine improvements have been bestowed upon it, this is
 *   no longer as clear as it may have once been.  Node.js Foundation
 *   releases 64-bit versions themselves, likely for good reason.
 *   Present-day operation of 64-bit binaries on 64-bit Windows
 *   platforms show clear performance benefits over their 32-bit
 *   siblings (e.g. 7-zip, et.al), so Meteor should also try to offer
 *   that same benefit by building and offering a 64-bit version.
 *   Meteor no longer supports Windows 32-bit.
 *
 * To be (more but far from completely) precise, the ABI for os.*
 * architectures includes a CPU type, a mode in which the code will be
 * run (eg, 64 bit), an executable file format (eg, ELF), a promise to
 * make any shared libraries available in a particular architecture,
 * and promise to set up the shared library search path
 * "appropriately". In the future it will also include some guarantees
 * about the directory layout in the environment, eg, location of a
 * directory where temporary files may be freely written. It does not
 * include any syscalls (beyond those used by code that customarily is
 * statically linked into every executable built on a platform, eg,
 * exit(2)). It does not guarantee the presence of any particular
 * shared libraries or programs (including any particular shell or
 * traditional tools like 'grep' or 'find').
 *
 * To model the shared libraries that are required on a system (and
 * the particular versions that are required), and to model
 * dependencies on command-line programs like 'bash' and 'grep', the
 * idea is to have a package named something like 'posix-base' that
 * rolls up a reasonable base environment (including such modern
 * niceties as libopenssl) and is supplied by the container. This
 * allows it to be versioned, unlike architectures, which we hope to
 * avoid versioning.
 *
 * Q: What does "x86" mean?
 * A: It refers to the traditional Intel architecture, which
 * originally surfaced in CPUs such as the 8086 and the 80386. Those
 * of us who are older should remember that the last time that Intel
 * used this branding was the 80486, introduced in 1989, and that
 * today, parts that use this architecture bear names like "Core",
 * "Atom", and "Phenom", with no "86" it sight. We use it in the
 * architecture name anyway because we don't want to depart too far
 * from Linux's architecture names.
 *
 * Q: Why do we call it "x86_32" instead of the customary "i386" or
 * "i686"?
 * A: We wanted to have one name for 32-bit and one name for 64-bit,
 * rather than several names for each that are virtual synonyms for
 * each (eg, x86_64 vs amd64 vs ia64, i386 vs i686 vs x86). For the
 * moment anyway, we're willing to adopt a "one size fits all"
 * attitude to get there (no ability to have separate builds for 80386
 * CPUs that don't support Pentium Pro extensions, for example --
 * you'll have to do runtime detection if you need that). And as long
 * as we have to pick a name, we wanted to pick one that was super
 * clear (it is not obvious to many people that "i686" means "32-bit
 * Intel", because why should it be?) and didn't imply too close of an
 * equivalence to the precise meanings that other platforms may assign
 * to some of these strings.
 */

// Valid architectures that Meteor officially supports.
export const VALID_ARCHITECTURES: Record<string, boolean> = {
  "os.osx.x86_64": true,
  "os.linux.x86_64": true,
  "os.windows.x86_64": true,
  "os.freebsd.x86_64": true,
};

// Returns the fully qualified arch of this host -- something like
// "os.linux.x86_32" or "os.osx.x86_64". Must be called inside
// a fiber. Throws an error if it's not a supported architecture.
//
// If you change this, also change scripts/admin/launch-meteor
let _host: string | null = null; // memoize

export function host() {
  if (!_host) {
    const run = function (...args: Array<string | boolean>) {
      const result = utils.execFileSync(args[0], args.slice(1)).stdout;

      if (! result) {
        throw new Error(`Can't get arch with ${args.join(" ")}?`);
      }

      return result.replace(/\s*$/, ''); // trailing whitespace
    };

    const platform = os.platform();

    if (platform === "darwin") {
      // Can't just test uname -m = x86_64, because Snow Leopard can
      // return other values.
      if (run('uname', '-p') !== "i386" ||
          run('sysctl', '-n', 'hw.cpu64bit_capable') !== "1") {
        throw new Error("Only 64-bit Intel processors are supported on OS X");
      }
      _host  = "os.osx.x86_64";
    } else if (platform === "linux") {
      const machine = run('uname', '-m');
      if (["x86_64", "amd64", "ia64"].includes(machine)) {
        _host = "os.linux.x86_64";
      } else {
        throw new Error(`Unsupported architecture: ${machine}`);
      }
    } else if (["linux", "freebsd", "openbsd", "netbsd"].includes(platform)) {
      var machine = run('uname', '-m');
      if (["i386", "i686", "x86"].includes(machine)) {
        _host = "os." + platform + ".x86_32";
      } else if (["x86_64", "amd64", "ia64"].includes(machine)) {
        _host = "os." + platform + ".x86_64";
      } else if (["armv6l"].includes(machine)) {
        _host = "os." + platform + ".armv6l";
      } else if (["armv7l"].includes(machine)) {
        _host = "os." + platform + ".armv7l";
      } else {
        throw new Error("Unsupported architecture: " + machine);
      }
    } else if (platform === "win32" && process.arch === "x64") {
        _host = "os.windows.x86_64";
    } else {
      throw new Error(`Unsupported operating system: ${platform}`);
    }
  }

  return _host;
}

// In order to springboard to earlier Meteor releases that did not have
// 64-bit Windows builds, Windows installations must be allowed to
// download 32-bit builds of meteor-tool.
export function acceptableMeteorToolArches(): string[] {
  if (os.platform() === "win32") {
    switch (utils.architecture()) {
    case "x86_32":
      return ["os.windows.x86_32"];
    case "x86_64":
      return [
        "os.windows.x86_64",
        "os.windows.x86_32",
      ];
    }
  }

  return [host()];
}

// 64-bit Windows machines that have been using a 32-bit version of Meteor
// are eligible to switch to 64-bit beginning with Meteor 1.6, which is
// the first version of Meteor that contains this code.
export function canSwitchTo64Bit(): boolean {
  // Automatically switching from 32-bit to 64-bit Windows builds is
  // disabled for the time being, since downloading additional builds of
  // meteor-tool isn't stable enough at the moment (on Windows, at least)
  // to introduce in a release candidate.
  return false &&
    utils.architecture() === "x86_64" &&
    host() === "os.windows.x86_32";
}

// True if `host` (an architecture name such as 'os.linux.x86_64') can run
// programs of architecture `program` (which might be something like 'os',
// 'os.linux', or 'os.linux.x86_64').
//
// `host` and `program` are just mnemonics -- `host` does not
// necessarily have to be a fully qualified architecture name. This
// function just checks to see if `program` describes a set of
// environments that is a (non-strict) superset of `host`.
export function matches(host: string, program: string): boolean {
  return host.substr(0, program.length) === program &&
    (host.length === program.length ||
     host.substr(program.length, 1) === ".");
}

const legacyArches = [
  "web.browser.legacy",
  // It's important to include web.browser.legacy resources in the Cordova
  // bundle, since Cordova bundles are built into the mobile application,
  // rather than being downloaded from a web server at runtime. This means
  // we can't distinguish between clients at runtime, so we have to use
  // code that works for all clients.
  "web.cordova",
];

export function isLegacyArch(arch: string): boolean {
  return legacyArches.some(la => matches(arch, la));
}

export function mapWhereToArches(where: string) {
  const arches: string[] = [];

  // Shorthands for common arch prefixes:
  // "server" => os.*
  // "client" => web.*
  // "legacy" => web.browser.legacy, web.cordova
  if (where === "server") {
    arches.push("os");
  } else if (where === "client") {
    arches.push("web");
  } else if (where === "modern") {
    arches.push("web.browser");
  } else if (where === "legacy") {
    arches.push(...legacyArches);
  } else {
    arches.push(where);
  }

  return arches;
}

// Like `supports`, but instead taken an array of possible
// architectures as its second argument. Returns the most specific
// match, or null if none match. Throws an error if `programs`
// contains exact duplicates.
export function mostSpecificMatch(host: string, programs: string[]): string | null  {
  let best: string | null = null;
  const seen: Record<string, boolean> = {};

  programs.forEach((program: string) => {
    if (seen[program]) {
      throw new Error(`Duplicate architecture: ${program}`);
    }

    seen[program] = true;

    if (matches(host, program) && (!best || program.length > best.length)) {
      best = program;
    }
  });

  return best;
}

// `programs` is a set of architectures (as an array of string, which
// may contain duplicates). Determine if there exists any architecture
// that is compatible with all of the architectures in the set. If so,
// returns the least specific such architecture. Otherwise (the
// architectures are disjoin) raise an exception.
//
// For example, for 'os' and 'os.osx', return 'os.osx'. For 'os' and
// 'os.linux.x86_64', return 'os.linux.x86_64'. For 'os' and 'browser', throw an
// exception.
export function leastSpecificDescription(programs: string[]): string {
  if (programs.length === 0) {
    return '';
  }

  // Find the longest string
  const longest = max(programs, (p: string) => p.length);

  // If everything else in the list is compatible with the longest,
  // then it must be the most specific, and if everything is
  // compatible with the most specific then it must be the least
  // specific compatible description.
  programs.forEach((program: string) => {
    if (!matches(longest, program)) {
      throw new Error(`Incompatible architectures: '${program}' and '${longest}'`);
    }
  });

  return longest;
}

export function withoutSpecificOs(arch: string): string {
  if (arch.substr(0, 3) === 'os.') {
    return 'os';
  }

  return arch;
}
